### PR TITLE
feat: implement organization context in the cli

### PR DESCRIPTION
**Pull request by Emyrk**
# What this does

The cli now uses the config files to store what organization it is currently referencing.

Adds `coder organizations current` to display the currently selected organization.

**This is a hidden command until we finish all related commands**

# Discussion

Currently I have this under `coder organizations { current | switch | ... }. Should we name this something else? Like `coder ctx current`, `coder ctx switch` etc? 

This could maybe include different deployments and logins if we do `ctx`.

## `coder organizations current`

Below is all the format options.

```
$ go install && coder organizations current                                                                                                                                
Current organization: admin (3736ca3e-840f-4473-88ad-7056f1a9d22a)

$ coder organizations show current --output=json
[
  {
    "id": "3736ca3e-840f-4473-88ad-7056f1a9d22a",
    "name": "admin",
    "created_at": "2024-02-21T23:22:42.667305Z",
    "updated_at": "2024-02-21T23:22:42.667305Z",
    "is_default": true
  }
]

$ coder organizations show current --output=table
ID                                    NAME   DEFAULT  
3736ca3e-840f-4473-88ad-7056f1a9d22a  admin  true   

# coder organizations show current --only-id
3736ca3e-840f-4473-88ad-7056f1a9d22a
```

### CODE REVIEW

File: cli/organization.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	"github.com/coder/coder/v2/cli/cliui"
+	"github.com/coder/coder/v2/codersdk"
+)
+
+func (r *RootCmd) organizations() *clibase.Cmd {
+	cmd := &clibase.Cmd{
+		Annotations: workspaceCommand,
+		Use:         "organizations { current }",
+		Short:       "Organization related commands",
+		Aliases:     []string{"organization", "org", "orgs"},
```
Comment (Emyrk) [inline cli/organization.go:16]: Should we instead do something like `coder ctx current`?

File: cli/organization.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+		Hidden:      true, // Hidden until these commands are complete.
```
Comment (Emyrk) [inline cli/organization.go:17]: I marked this as hidden, I think we can defer the actual command name discussion and get the code in. Don't want to bikeshed and prevent code progress.

File: cli/organization.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+				}
+				return fmt.Sprintf("Current organization: %s (%s)\n", typed[0].Name, typed[0].ID.String()), nil
+			}),
+			cliui.TableFormat([]codersdk.Organization{}, []string{"id", "name", "default"}),
+			cliui.JSONFormat(),
+		)
+		onlyID = false
+	)
+	cmd := &clibase.Cmd{
+		Use:   "current",
```
Comment (mafredri) [inline cli/organization.go:51]: Suggestion: `s/current/show/`.

This matches better what we have in the cli already, `coder show <workspace>`, `coder users show <me>`, and now `coder org show [current|me|uuid]`. I'm thinking the arg can be optional.

File: cli/organization.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+		Short: "Show the current selected organization the cli will use.",
```
Comment (mafredri) [inline cli/organization.go:52]: If rename goes through, this text could read something like:

```suggestion
		Short: "Show the organization, if no argument is given, the organization currently in use will be shown.",
```

File: cli/organization.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+		Middleware: clibase.Chain(
+			r.InitClient(client),
+		),
+		Options: clibase.OptionSet{
+			{
+				Name:        "only-id",
```
Comment (mafredri) [inline cli/organization.go:58]: This could be achieved by `-c id` as well, although the custom formatter would need to handle it.

File: cli/root.go [hunk 6]
```diff
@@ -698,14 +699,44 @@ func (r *RootCmd) createAgentClient() (*agentsdk.Client, error) {
...
+		}
+		return orgs[index], nil
+	}
+
+	// User did not select an organization, so use the default.
+	index := slices.IndexFunc(orgs, func(org codersdk.Organization) bool {
+		return org.IsDefault
+	})
+	if index < 0 {
+		return codersdk.Organization{}, xerrors.Errorf("unable to determine current organization. Use 'coder switch <org>' to select an organization to use")
```
Comment (mafredri) [inline cli/root.go:736]: Should this be `coder org switch` instead? Also, should we include this help in the switch PR instead?

File: cli/templatecreate_test.go [hunk 7]
```diff
@@ -243,19 +243,7 @@ func TestTemplateCreate(t *testing.T) {
...
-			{match: "Upload", write: "yes"},
-		}
-		for _, m := range matches {
-			pty.ExpectMatch(m.match)
-			if len(m.write) > 0 {
-				pty.WriteLine(m.write)
-			}
-		}
-
+		pty.ExpectMatch("context canceled")
```
Comment (mafredri) [inline cli/templatecreate_test.go:246]: Why this change?

File: cli/organization.go (hunk 1 above)
Comment (mafredri) [reply cli/organization.go]: I'd be interested in discussing this further, ultimately I'd like for us to support config files for the CLI too, you might not want same behavior across all orgs. Having this support those use-cases as well would be nice.

File: cli/root.go [hunk 8]
```diff
@@ -698,14 +699,44 @@ func (r *RootCmd) createAgentClient() (*agentsdk.Client, error) {
...
+			return codersdk.Organization{}, fmt.Errorf("read selected organization from config file %q: %w", conf.Organization(), err)
+		}
+		selected = org
+	}
+
+	// Verify the org exists and the user is a member
 	orgs, err := client.OrganizationsByUser(inv.Context(), codersdk.Me)
 	if err != nil {
-		return codersdk.Organization{}, nil
+		return codersdk.Organization{}, err
```
Comment (Emyrk) [inline cli/root.go:716]: @mafredri The unit test expecting `context cancelled` had to be changed because of this change. (https://github.com/coder/coder/pull/12259#discussion_r1500825567)

Before, if you failed to fetch the orgs, we'd return a nil uuid. I made this api failure fail the command now. So the unit tests fails differently now.

File: cli/organization.go (hunk 3 above)
Comment (Emyrk) [reply cli/organization.go:51]: That is a good idea. If no arg is provided, I guess I could default to `current`?

File: cli/root.go (hunk 8 above)
Comment (mafredri) [reply cli/root.go:716]: I wouldn't expect `TestTemplateCreate/WithVariablesFileWithoutRequiredValue` to be testing for org failure though 🤔

File: cli/organization.go (hunk 3 above)
Comment (mafredri) [reply cli/organization.go:51]: Yep, that sounds good 👍🏻

Comment (Emyrk) [reply cli/organization.go]: All the output formats work as expected.

```
$ coder organizations show me
Organizations you are a member of:
        admin (653e414b-9eb7-41cb-b11c-38cf90b29db3)
        steven-org (9616b384-38bb-4284-861b-8db4fc337d6d)
        bia-org (75e6295c-0bfb-4195-8898-b9388879dafb)
```

File: cli/organization.go (hunk 5 above)
Comment (Emyrk) [reply cli/organization.go:58]: You mean a column selector `-c`? Unfortunately that still prints the headers though for the table.

I imagine this to be used in a script, and you want to quickly just get the `ID` to place in a request.

File: cli/organization.go (hunk 1 above)
Comment (Emyrk) [reply cli/organization.go]: Good point. This command is marked as hidden, so we can tweak it before we release.

Multi orgs is still a bit out, but this is going to be really helpful in development.

File: cli/root.go (hunk 8 above)
Comment (Emyrk) [reply cli/root.go:716]: I didn't write the test, but there is a comment indicating this cli error is expected. We cancel the context before running the command.

https://github.com/coder/coder/blob/98666fd7d7d3e3449b4427d5f5ab6b8e8958765a/cli/templatecreate_test.go#L238-L244

Comment (mafredri) [reply cli/root.go:716]: If we're testing against a cancelled context, that test is useless. I think it should either be removed or fixed. 😅 It's definitely not testing what the test name is indicating.

**Review by mafredri — APPROVED** [review]
I'm not happy with that test that tests for context cancellation, but not letting that block this PR, so approved. 👍🏻

**Comment by Emyrk** [comment]
> I'm not happy with that test that tests for context cancellation, but not letting that block this PR, so approved. 👍🏻

Yea, I'm not 100% sure what the test is intended to do because the prior behavior makes little sense to me as well.
